### ISSUE TITLE
LineageOS v21.0 (Android 14) Error: libart.so: unable to find export '_ZNK3art2gc4Heap15MayUseCollectorENS0_13CollectorTypeE

### ISSUE THREAD

**Issue by FanchangWang** [frida/frida#2925]
### Device information：

- Redmi K30 Pro (lmi)
- LineageOS v21.0 (Android 14)
- frida-server-16.4.7-android-arm64

### 
```
 ./frida-server-16.4.7-android-arm64
{
	"type":"error","description":"Error: libart.so: unable to find export '_ZNK3art2gc4Heap15MayUseCollectorENS0_13CollectorTypeE'","stack":"Error: libart.so: unable to find export '_ZNK3art2gc4Heap15MayUseCollectorENS0_13CollectorTypeE'\n    
	at Function.value [as getExportByName] 	(frida/runtime/core.js:250:1)\n    
	at Ft (frida/node_modules/frida-java-bridge/lib/android.js:837:1)\n    
	at Tt (frida/node_modules/frida-java-bridge/lib/android.js:762:1)\n    
	at mn.replace (frida/node_modules/frida-java-bridge/lib/android.js:1331:1)\n    
	at Function.set [as implementation] (frida/node_modules/frida-java-bridge/lib/class-factory.js:1189:1)\n    
	at Function.set [as implementation] (frida/node_modules/frida-java-bridge/lib/class-factory.js:1103:1)\n    
	at installLaunchTimeoutRemovalInstrumentation (/internal-agent.js:424:24)\n    
	at init (/internal-agent.js:51:3)\n    
	at c.perform (frida/node_modules/frida-java-bridge/lib/vm.js:12:1)\n    
	at _performPendingVmOps (frida/node_modules/frida-java-bridge/index.js:250:1)",
	"fileName":"frida/runtime/core.js","lineNumber":250,"columnNumber":1
}
```

**Comment by poldenais** [comment]
Im getting a similar issue after updating frida and trying to run a script on an android Emulator.

```
Spawned `com.APPNAME`. Resuming main thread!
Error: libart.so: unable to find export '_ZNK3art2gc4Heap15MayUseCollectorENS0_13CollectorTypeE'
    at value (frida/runtime/core.js:250)
    at Ft (frida/node_modules/frida-java-bridge/lib/android.js:837)
    at Tt (frida/node_modules/frida-java-bridge/lib/android.js:762)
    at replace (frida/node_modules/frida-java-bridge/lib/android.js:1331)
    at set (frida/node_modules/frida-java-bridge/lib/class-factory.js:1189)
    at set (frida/node_modules/frida-java-bridge/lib/class-factory.js:1103)
    at <anonymous> (frida/node_modules/frida-java-bridge/index.js:224)
    at <anonymous> (frida/node_modules/frida-java-bridge/lib/vm.js:12)
    at _performPendingVmOpsWhenReady (frida/node_modules/frida-java-bridge/index.js:244)
    at perform (frida/node_modules/frida-java-bridge/index.js:204)
    at <eval> (/Users/username/Desktop/Frida/fingerprint-bypass.js:21)
```

**Comment by JorySeverijnse** [comment]
I am having the same issue on lineageos 21 oneplus 9 pro rooted

**Comment by unf0rgvn** [comment]
I'm facing the same issue

**Comment by unf0rgvn** [comment]
Found a solution. If you're using Android Studio. Use the API 34, instead 35, then you may run it without erros.

**Comment by tothi** [comment]
how to build it using Android Studio? building it from cmdline (configure+make), and able to customize only the Android NDK root folder, and if it is not r25, it does not build. so the question is, how to specify API level?

**Comment by rhysperry111** [comment]
Also having this problem. Had a crack at recompiling myself but I ran into the same problem @tothi did

```
ANDROID_NDK_ROOT=/tmp/android-ndk-r27 ANDROID_NDK_VERSIONS=r27 ./configure --host=android-arm64
Downloading SDK 20240729 for android-arm64...
Extracting SDK...
NDK r25 is required (found r27, which is unsupported)
```

**Comment by zb3** [comment]
Here's the build of frida-server with these GC-related methods removed.. it is probably unstable and incorrect (but works for my case - better than nothing). Build sent by my trusted friend Jia Tan, so should be safe to use.

[frida-server-android-arm64.zip](https://github.com/user-attachments/files/16573771/frida-server-android-arm64.zip)

**Comment by unf0rgvn** [comment]
> how to build it using Android Studio? building it from cmdline (configure+make), and able to customize only the Android NDK root folder, and if it is not r25, it does not build. so the question is, how to specify API level?

I really don't know how to do it in cmdline. I made the installation in GUI Android Studio:
![github](https://github.com/user-attachments/assets/32433acb-6918-48ee-a09e-55a35c0a42e6)

**Comment by quantum2274** [comment]
> unf0rgvn

this worked for me too

**Comment by NgoHuy** [comment]
the PR https://github.com/frida/frida-java-bridge/pull/330 and https://github.com/frida/frida-java-bridge/pull/332 fix this issue, please test it on your end by rebuild the package.

**Comment by zb3** [comment]
@NgoHuy, have you seen my compiled binary in the comment above? For me frida worked, but failed to intercept some functions.

**Comment by tothi** [comment]
@NgoHuy still having this issue (after applying both of the PRs:

```
{"type":"error","description":"Error: libart.so: unable to find export '_ZN3art2gc9collector17ConcurrentCopying12CopyingPhaseEv'","stack":"Error: libart.so: unable to find export '_ZN3art2gc9collector17ConcurrentCopying12CopyingPhaseEv'\n    at Function.value [as getExportByName] (frida/runtime/core.js:250:1)\n    at Ft (frida/node_modules/frida-java-bridge/lib/android.js:843:1)\n    at Tt (frida/node_modules/frida-java-bridge/lib/android.js:762:1)\n    at mn.replace (frida/node_modules/frida-java-bridge/lib/android.js:1335:1)\n    at Function.set [as implementation] (frida/node_modules/frida-java-bridge/lib/class-factory.js:1189:1)\n    at Function.set [as implementation] (frida/node_modules/frida-java-bridge/lib/class-factory.js:1103:1)\n    at installLaunchTimeoutRemovalInstrumentation (/internal-agent.js:424:24)\n    at init (/internal-agent.js:51:3)\n    at c.perform (frida/node_modules/frida-java-bridge/lib/vm.js:12:1)\n    at _performPendingVmOps (frida/node_modules/frida-java-bridge/index.js:250:1)","fileName":"frida/runtime/core.js","lineNumber":250,"columnNumber":1}
```

**Comment by NgoHuy** [comment]
You need to run make clean and make and again in build directory.

**Comment by NgoHuy** [comment]
You should use this PR for testing, it includes all fixes for this issue

**Comment by tothi** [comment]
@NgoHuy Thanks! At last it seems to be working (not saying it is stable, but at least it is running), used the all-in-one PR from Thinh:

https://github.com/frida/frida-java-bridge/pull/330

The build process was a little bit tricky, because the fix is in frida-java-bridge what is an npm dependency of the frida package. By default the frida build process pulls the required npm packages from the official public npm repo (called "registry" in npm terminology).

So here is how I managed to build it (until the PR gets approved):

1. Fire up a local npm registry what proxies everything to an upstream official, but serves packages from the local storage if it is published locally. I used 'verdaccio': https://verdaccio.org/

After launching verdaccio (using the default config should work) we need to set the npm registry to verdaccio:

`npm set registry http://localhost:4873/`

Now every npm dependency should be served by verdaccio, 1st from the local storage, if it is missing, it proxies the request to the default public upstream.

Authentication might be needed to set up correctly for the default config (look for `htpasswd`, `npm login`, etc.).

2. Build the frida-java-bridge with the PR checked out (only 330 is required) and publish it to the local npm registry:

```
git clone https://github.com/frida/frida-java-bridge && cd frida-java-bride
git pull origin pull/330/head
npm publish
```

If `npm publish` returns 409 conflict, you may need `npm unpublish` first.

3. Now building frida should pick up the patched frida-java-bridge.

```
git clone https://github.com/frida/frida && cd frida
export ANDROID_NDK_ROOT=/opt/android-sdk/ndk/25.2.9519653/  # need to point to your Android NDK folder (r25 is required)
./configure --host=android-arm64
make
```

The `frida-server` binary built for android-arm64 should be at `build/subprojects/frida-core/server/frida-server`.

**Comment by NgoHuy** [comment]
> @NgoHuy Thanks! At last it seems to be working (not saying it is stable, but at least it is running), used the all-in-one PR from Thinh:
> 
> [frida/frida-java-bridge#330](https://github.com/frida/frida-java-bridge/pull/330)
> 
> The build process was a little bit tricky, because the fix is in frida-java-bridge what is an npm dependency of the frida package. By default the frida build process pulls the required npm packages from the official public npm repo (called "registry" in npm terminology).
> 
> So here is how I managed to build it (until the PR gets approved):
> 
>     1. Fire up a local npm registry what proxies everything to an upstream official, but serves packages from the local storage if it is published locally. I used 'verdaccio': https://verdaccio.org/
> 
> 
> After launching verdaccio (using the default config should work) we need to set the npm registry to verdaccio:
> 
> `npm set registry http://localhost:4873/`
> 
> Now every npm dependency should be served by verdaccio, 1st from the local storage, if it is missing, it proxies the request to the default public upstream.
> 
> Authentication might be needed to set up correctly for the default config (look for `htpasswd`, `npm login`, etc.).
> 
>     2. Build the frida-java-bridge with the PR checked out (only 330 is required) and publish it to the local npm registry:
> 
> 
> ```
> git clone https://github.com/frida/frida-java-bridge && cd frida-java-bride
> git pull origin pull/330/head
> npm publish
> ```
> 
> If `npm publish` returns 409 conflict, you may need `npm unpublish` first.
> 
>     3. Now building frida should pick up the patched frida-java-bridge.
> 
> 
> ```
> git clone https://github.com/frida/frida && cd frida
> export ANDROID_NDK_ROOT=/opt/android-sdk/ndk/25.2.9519653/  # need to point to your Android NDK folder (r25 is required)
> ./configure --host=android-arm64
> make
> ```
> 
> The `frida-server` binary built for android-arm64 should be at `build/subprojects/frida-core/server/frida-server`.

You should replace the file in build/subprojects/frida-gum/bindings/runtime.xx.p/node_modules/frida-java-bridge/lib/android.js then run `make clean && make` again

**Comment by rhysperry111** [comment]
Are we just waiting on a dependency version bump + new release?

**Comment by devnoname120** [comment]
Duplicate of: https://github.com/frida/frida/issues/2851

**Comment by zcfrank1st** [comment]
When to release the fixed version？ 16.5.2 got the error, too

**Comment by strva-re** [comment]
got the error, too
